### ISSUE TITLE
Slow Performance of ProductMetadata::getVersion

### ISSUE THREAD

**Issue by beberlei** [magento/magento2#24025]
### Preconditions (*)

Any Magento 2 version.

### Steps to reproduce (*)

1. Have a more complex composer.json with source checked repositories (git checkouts for example).
2. A plugin that calls `ProductMetadata::getVersion` to check for Magento version. 

### Expected result (*)

Calling ProductMetadata::getVersion, I expected it  to be a function returning a constant. Super-fast.

### Actual result (*)

147ms in this case with 7 calls to proc_open, calling svn, git, hg to retrieve versions of source packages. In addition the PHP based JSON Parser is triggered for ~15ms. This overhead is present regardless f all packages were dist or source based and their version were known.

This output is from Tideways callgraph profiler:

![tideways_magento2_getversion](https://user-images.githubusercontent.com/26936/62524105-5c21b080-b835-11e9-9625-8e3549f47f4d.png)

This is related to the closed #23242 ticket that wasn't reproduced.

This performance issue is especially critical, because they way plugins check the magento version, this overhead is usually present in *all* requests to a Magento 2 shop.

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @davidverholen 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-18958` were created

**Issue Available**: @davidverholen, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by davidverholen** [comment]
it has been a constant originally. It should probably be changed back again with the constant update being part of the release process
https://github.com/magento/magento2/commit/e46a3cfeccae6f8fcd1eb408719f48d2e039c433#diff-162b082e41b9d0990e5dcc7439d19f26

This method is called several times in the core. Also, 3rd party extensions are using it to check compatibility of features.
In very many cases this leads to every uncached request being 100ms - 200ms slower

**Comment by beberlei** [comment]
The easiest fix would be using the composer json finder there and parse the versions out intead of using `ComposerInformation`:

```patch
diff --git a/lib/internal/Magento/Framework/App/ProductMetadata.php b/lib/internal/Magento/Framework/App/ProductMetadata.php
index c9fde94352a..5fdef2ba5c5 100644
--- a/lib/internal/Magento/Framework/App/ProductMetadata.php
+++ b/lib/internal/Magento/Framework/App/ProductMetadata.php
@@ -7,10 +7,7 @@
  */
 namespace Magento\Framework\App;
 
-use Magento\Framework\Composer\ComposerFactory;
 use \Magento\Framework\Composer\ComposerJsonFinder;
-use \Magento\Framework\App\Filesystem\DirectoryList;
-use \Magento\Framework\Composer\ComposerInformation;
 
 /**
  * Class ProductMetadata
@@ -41,11 +38,6 @@ class ProductMetadata implements ProductMetadataInterface
      */
     protected $composerJsonFinder;
 
-    /**
-     * @var \Magento\Framework\Composer\ComposerInformation
-     */
-    private $composerInformation;
-
     /**
      * @param ComposerJsonFinder $composerJsonFinder
      */
@@ -62,11 +54,13 @@ class ProductMetadata implements ProductMetadataInterface
     public function getVersion()
     {
         if (!$this->version) {
-            if (!($this->version = $this->getSystemPackageVersion())) {
-                if ($this->getComposerInformation()->isMagentoRoot()) {
-                    $this->version = $this->getComposerInformation()->getRootPackage()->getPrettyVersion();
-                } else {
-                    $this->version = 'UNKNOWN';
+            $composerJsonFile = $this->composerJsonFinder->findComposerJson();
+            $installedPackages = json_decode(file_get_contents(dirname($composerJsonFile) . "/vendor/composer/installed.json"));
+
+            foreach ($installedPackages as $package) {
+                if (in_array($package['name'], ['magento/product-community-edition', 'magento/product-enterprise-edition'])) {
+                    $this->version = $package['version'];
+                    break;
                 }
             }
         }
@@ -92,37 +86,4 @@ class ProductMetadata implements ProductMetadataInterface
     {
         return self::PRODUCT_NAME;
     }
-
-    /**
-     * Get version from system package
-     *
-     * @return string
-     * @deprecated 100.1.0
-     */
-    private function getSystemPackageVersion()
-    {
-        $packages = $this->getComposerInformation()->getSystemPackages();
-        foreach ($packages as $package) {
-            if (isset($package['name']) && isset($package['version'])) {
-                return $package['version'];
-            }
-        }
-        return '';
-    }
-
-    /**
-     * Load composerInformation
-     *
-     * @return ComposerInformation
-     * @deprecated 100.1.0
-     */
-    private function getComposerInformation()
-    {
-        if (!$this->composerInformation) {
-            $directoryList              = new DirectoryList(BP);
-            $composerFactory            = new ComposerFactory($directoryList, $this->composerJsonFinder);
-            $this->composerInformation  = new ComposerInformation($composerFactory);
-        }
-        return $this->composerInformation;
-    }
 }
```

**Comment by beberlei** [comment]
A better approach (but more complex code wise) would probably be to write a `magento-version.php` file into the cache directory upon cache generation that contains the version:

```php
<?php
// cache/magento-version.php
return '2.3.2';
```

and then doing `return require $this->cacheDir . '/magento-version.php';` in `getVersion()` instead.

**Comment by sprankhub** [comment]
There was also PR #19133, which introduced caching for the `getVersion` method. However, it was rejected.

**Comment by davidverholen** [comment]
while the core might not really need it (because it knows its own version) it is a valid use case for 3rd Party developers for not having to maintain different code bases for each magento version.

Even when it's not considered best practice, still, many 3rd party extensions use it and so this is a very real problem

**Comment by magento-engcom-team** [comment]
Hi @beberlei. Thank you for your report. 
The issue has been fixed in magento/magento2#24030 by @davidverholen in 2.3-develop branch
Related commit(s):
 - [f0856e8575bc0cc86543d14da2bd45d8e457fa25](https://github.com/magento/magento2/commit/f0856e8575bc0cc86543d14da2bd45d8e457fa25)
 - [ed1fb0358045fd988c59a00e68099cf3632e5865](https://github.com/magento/magento2/commit/ed1fb0358045fd988c59a00e68099cf3632e5865)
 - [b2ca7c878dbdc27b69ddd34646f84f0dc275d073](https://github.com/magento/magento2/commit/b2ca7c878dbdc27b69ddd34646f84f0dc275d073)
 - [b397814fefdb1d4ae35c8adf7c108bac96e90ffa](https://github.com/magento/magento2/commit/b397814fefdb1d4ae35c8adf7c108bac96e90ffa)

The fix will be available with the upcoming 2.3.4 release.

**Comment by erikhansen** [comment]
For the record, I encountered this issue on a Magento OS 2.3.2 site, as several Amasty extensions were calling `\Magento\Framework\App\ProductMetadata::getVersion`. After applying #24030 as [a patch](https://patch-diff.githubusercontent.com/raw/magento/magento2/pull/24030.patch) (via [composer-patches](https://www.classyllama.com/blog/create-apply-patches-magento-2)), the ~100ms slowdown went away.

**Comment by VladimirZaets** [comment]
Hi @beberlei. Thank you for your report. 
The issue has been fixed in magento/magento2#26001 by @luklewluk in 2.4-develop branch
Related commit(s):
 - [16ab63b347dd4628ed1cc1947929ff44b6bad90c](https://github.com/magento/magento2/commit/16ab63b347dd4628ed1cc1947929ff44b6bad90c)
 - [ab6cabd682ae18a48aaf1429232047d90e4e881c](https://github.com/magento/magento2/commit/ab6cabd682ae18a48aaf1429232047d90e4e881c)
 - [0b7c34ef79307da52e088b6d3652663e9cd2c3a1](https://github.com/magento/magento2/commit/0b7c34ef79307da52e088b6d3652663e9cd2c3a1)

The fix will be available with the upcoming 2.4.0 release.

**Comment by erikhansen** [comment]
After upgrading from Magento 2.3.2 to 2.3.3, [the patch I mentioned above](https://github.com/magento/magento2/issues/24025#issuecomment-542476510) no longer applies (since the underlying code changed). If you're using 2.3.3 and you need to fix this issue, you can use this updated patch (it contains the relevant changes from [16ab63](https://github.com/magento/magento2/commit/16ab63b347dd4628ed1cc1947929ff44b6bad90c)):

[magento-framework-github-24025-fix-composer-version-lookup.patch.txt](https://github.com/magento/magento2/files/4112454/magento-framework-github-24025-fix-composer-version-lookup.patch.txt) (remove the `.txt` from the filename)
